rewrite-java/src/main/java/org/openrewrite/java/JavaSourceVisitor.java

```diff
@@ -45,32 +46,64 @@
+     * <P><P>
+     * <li>A static class declaration</li>
+     * <li>An enumeration declaration</li>
+     * <li>An interface declaration</li>
+     * <li>An annotation declaration</li>
+     *
+     * @param base A pointer within the AST that is used to establish the "base lexical scope".
+     * @param child A pointer within the AST that will be traversed (up the tree) looking for an intersection with the base lexical scope.
+     * @return true if the child is in within the lexical scope of the base
+     */
+    default boolean isInSameNameScope(Cursor base, Cursor child) {
+        //First establish the base scope by finding the first enclosing element.
+        Tree baseScope = base.getPathAsStream()
                 .filter(t -> t instanceof J.Block ||
                         t instanceof J.MethodDecl ||
                         t instanceof J.Try ||
                         t instanceof J.ForLoop ||
-                        t instanceof J.ForEachLoop).findAny()
-                .map(higherNameScope -> lower.getPathAsStream()
-                        .filter(t -> {
-                            takeWhile.set(takeWhile.get() && (
-                                    !(t instanceof J.ClassDecl) ||
-                                            (((J.ClassDecl) t).getKind() instanceof J.ClassDecl.Kind.Class &&
-                                                    !((J.ClassDecl) t).hasModifier("static"))));
-                            return takeWhile.get();
-                        })
-                        .anyMatch(higherNameScope::equals))
-                .orElse(false);
+                        t instanceof J.ForEachLoop)
+                .findFirst().orElseThrow(() -> new IllegalArgumentException("The base cursor does not have an scoped context."));
+
+        //Now walk up the child path looking for the base scope.
+        for (Iterator<Tree> it = child.getPath(); it.hasNext(); ) {
+            Tree childScope = it.next();
+            if (childScope instanceof J.ClassDecl) {
+                J.ClassDecl childClass = (J.ClassDecl) childScope;
+                if (!(childClass.getKind() instanceof J.ClassDecl.Kind.Class) || childClass.hasModifier("static")) {
+                    //Short circuit the search if a terminating element is encountered.
+                    break;
+                }
+            }
+            if (baseScope.isScope(childScope)) {
+                return true;
+            }
+        }
+
+        return false;
     }
 
     /**
-     * @param lower The cursor of the lower scoped tree element to check.
-     * @return Whether this cursor shares the same name scope as {@code scope}.
+     * This method determines if the element referenced by the child cursor has visibility to variables and declarations
+     * that are visible to the element referenced by the current cursor. (Is the child within the lexical scope of the base?)
+     *
+     * See {@link JavaSourceVisitor#isInSameNameScope}
```

Comment:
I almost got caught in an infinite loop reading this same comment over and over 😉